virtwho/manager/subscriptionmanager/subscriptionmanager.py

```diff
@@ -220,13 +205,64 @@ def hypervisorCheckIn(self, report, options=None):
             if hasattr(e, 'code'):
                 raise ManagerError("Communication with subscription manager failed with code %d: %s" % (e.code, str(e)))
             raise ManagerError("Communication with subscription manager failed: %s" % str(e))
+
         if is_async is True:
             report.state = AbstractVirtReport.STATE_PROCESSING
             report.job_id = result['id']
         else:
             report.state = AbstractVirtReport.STATE_FINISHED
         return result
 
+    def is_server_async(self, report, connection=None):
+        """
+        Check if server has capability 'hypervisor_async'.
+        """
+        if connection is None:
+            self._connect(report.config)
+
+        self.logger.debug("Checking if server has capability 'hypervisor_async'")
+        is_async = hasattr(self.connection, 'has_capability') and self.connection.has_capability('hypervisors_async')
+
+        if is_async:
+            self.logger.debug("Server has capability 'hypervisors_async'")
+        else:
+            self.logger.debug("Server does not have 'hypervisors_async' capability")
+
+        if is_async and os.environ.get('VIRTWHO_DISABLE_ASYNC', '').lower() in ['1', 'yes', 'true']:
+            self.logger.info("Async reports are supported but explicitly disabled")
+            is_async = False
+
+        return is_async
+
+    def hypervisor_mapping(self, report, is_async, connection=None):
+        """
+        Return mapping of hypervisor
+        """
+        if connection is None:
+            self._connect(report.config)
+
+        mapping = report.association
+        serialized_mapping = {}
+
+        if is_async:
+            # Transform the mapping into the async version
+            serialized_mapping = {'hypervisors': [h.toDict() for h in mapping['hypervisors']]}
+        else:
+            # Reformat the data from the mapping to make it fit with
+            # the old api.
+            for hypervisor in mapping['hypervisors']:
+                guests = [g.toDict() for g in hypervisor.guestIds]
+                serialized_mapping[hypervisor.hypervisorId] = guests
+
+        hypervisor_count = len(mapping['hypervisors'])
+        guest_count = sum(len(hypervisor.guestIds) for hypervisor in mapping['hypervisors'])
+
+        self.logger.info('Hosts-to-guests mapping for config "%s": %d hypervisors and %d guests found',
```

Comment:
I think this log message should be output for each report gathered from each source. In other words, this should probably be output when the reports are gathered and not when they are being sent.

So if there is a configuration "libvirt_1" and a configuration "libvirt_2", I think we should see this message both times. Once for the info gathered from "libvirt_1" and another time for the info gathered from "libvirt_2".